### ISSUE TITLE
3 questions from running this online tutorial

### ISSUE THREAD

**Issue by jielab** [jlmelville/uwot#79]
Dear James:

The tutorial says that "f**or high dimensional datasets (> 100-1000 columns) using PCA to reduce dimensionality is highly recommended to avoid the nearest neighbor search taking a long time. Keeping only 50 dimensions can speed up calculations without affecting the visualization much**". I asked a similar question before. For the SARS-COV-2 dataset that has ~30,000 genotype columns, so, should I run PCA first to reduce the data to <50 columns? 

For the section "“**# Add new points to an existing embedding**", what exactly does "**embedding**" mean here? I think literally "embedding" means "attaching". Does it mean that we first generate UMAP on a "train" dataset and then "embed" the UMAP of some "test" dataset into the same plot as teh "train" dataset? This way, we don't need to run UMAP again and again for **train+test1**, for **trait+test2**, etc., Correct? In PCA anaylysis this is usually called "**projection**".

I successfully copied and run all scripts in the online tutorial, except that:
1. for line "**mnist_nn_spca <- umap(mnist, nn_method = mnist_nn$nn, init = spca)**", it could NOT find "spca".
2.  Where is the script to generate the plot like below? I did not see it.

![mnist-py](https://user-images.githubusercontent.com/26947455/127662378-0257ff9a-09bd-4067-adc3-ae9ba1accad8.png)

Thank you veyr much & best regadrs,
Jie

**Comment by jlmelville** [comment]
> The tutorial says that "f**or high dimensional datasets (> 100-1000 columns) using PCA to reduce dimensionality is highly recommended to avoid the nearest neighbor search taking a long time. Keeping only 50 dimensions can speed up calculations without affecting the visualization much**". I asked a similar question before. For the SARS-COV-2 dataset that has ~30,000 genotype columns, so, should I run PCA first to reduce the data to <50 columns?

The difference between this and lmcinnes/umap#732 is that the nearest neighbor search used in uwot can take a very long time to return accurate results for datasets with a large number of columns, so reducing the dimensionality before running `uwot::umap` may be necessary. That is just an unfortunate requirement for the use of the `uwot` package, it's not inherent to UMAP itself. If you use the `umap-learn` Python package you would not need to do this.

The reason I don't think you should always run PCA on your data before input to UMAP is because of the situation you are now in: you are going to use PCA but how many components should you keep? The answer to that is: I don't know. No one knows. You will need to come up with some way to satisfy yourself that you haven't removed useful information from the dataset. I can't guarantee that 50 components is the right number. That works out ok for `mnist`, but may not be suitable for other datasets. Also note that that extracting a large number of components via PCA can also take a very long time.

> For the section "“**# Add new points to an existing embedding**", what exactly does "**embedding**" mean here?

It means the coordinates in the low-dimensional space from running UMAP.

> In PCA anaylysis this is usually called "**projection**".

Yes, it's like that. You can run UMAP on the training data once, then project the test set data using the returned UMAP "model".

> 1. for line "**mnist_nn_spca <- umap(mnist, nn_method = mnist_nn$nn, init = spca)**", it could NOT find "spca".

This is a bug, it should say: `mnist_nn_spca <- umap(mnist, nn_method = mnist_nn$nn, init = "spca")` (with `spca` as a string). I have fixed this, thank you for reporting it.

> 2. Where is the script to generate the plot like below? I did not see it.

`uwot` does not come with any plotting functions. You can use whatever R plotting library you like. For generating these sorts of plots specifically, I wrote a package you can find at <https://github.com/jlmelville/vizier>. After installation (e.g. via `devtools::install_github("jlmelville/vizier"`), then assuming the output of `uwot::umap` is in `mnist_umap`, you can make a plot via:

```R
vizier::embed_plot(coords = mnist_umap, x = mnist$Label, cex = 0.1, alpha_scale = 0.05)
```

**Comment by jielab** [comment]
All cear now. Thank you very much, James!

Best regadrs,
jie

**Comment by jielab** [comment]
One minor thing, I feel that the plot shown on your vizier page is NOT as impressive as that shown in the UMAP github tutorial.

For example, a plot like below could not be generated by your vizier R package, correct?
If so, it would be wonderful to enhance the plotting capacity of vizier.

![image](https://user-images.githubusercontent.com/26947455/127711067-4087cd56-90a7-4319-b969-39544a3e049a.png)


Best regards,
Jie

**Comment by jlmelville** [comment]
> One minor thing, I feel that the plot shown on your vizier page is NOT as impressive as that shown in the UMAP github tutorial.

Agreed.

> For example, a plot like below could not be generated by your vizier R package, correct?

Correct. And it never will be possible as I have no plans to expand the capacity of the package. It's simply meant to make it easy to produce an ok plot.

> If so, it would be wonderful to enhance the plotting capacity of vizier.

The package isn't intended to produce publication-ready images. You will have to look at a more specialized package for that.
